### ISSUE TITLE
[Ready for Payment][$1000] iOS - Chat - "Edit comment" input field scrolls up when tap enter in 1st line of text

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#19507]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Precondition - User sent a text message in chat (e.g. 1 word or 1 line of text)
1. Open IOS app
2. Open chat from the precondition
3. Long press the message
4. Tap "Edit comment"
5. Add 6 lines of text below the 1st line so there is a total of 7 lines
6. Put the cursor at the end of 1st line
7. Tap "Enter"
8. Place cursor at the end of 7th line / on the 8th line
9, Place cursor back at the end of the 1st line
10. Tap "Enter"

## Expected Result:
1st line of text hides above and cursor is focus on the next line of text and visible

## Actual Result:
At step 7 the comment field scrolls slightly up and cursor also scrolls out of the view.
At step 10 field focus scrolls back to line 7/8

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [x] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.17.0

**Reproducible in staging?:** yes
 
**Reproducible in production?:** yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://platform.applause.com/services/links/v1/external/197b3277eef42e2b057288d9b61067c2fce3fc4a98ffe01c313143ea19511833

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0113571ce6c5a8b04e</li>
        <li>Upwork Job ID: 1664050603930722304
<li>Last Price Increase: 2024-08-08</li></li>
        2023-07-19<li>Automatic offers: </li>
<ul>
<li> |  | 0</li>
<li>fabriziobertoglio1987 | Contributor | 25744429</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @alexpensify</details>

**Comment by alexpensify** [comment]
I was assigned when I was offline, I'm still catching up from being OOO. I'll test tomorrow.

**Comment by alexpensify** [comment]
@kbecciv - to confirm, is this bug within the native app or when using Safari on the iPhone? These two points need to be clarified (see screenshot). I tried testing but need to confirm if I need to be using the iPhone application or the app via a mobile browser. Thanks!  

![image](https://github.com/Expensify/App/assets/8107070/d42dd1a5-a8eb-490d-a63a-f345dd1d6c23)

**Comment by alexpensify** [comment]
Waiting for the updated testing steps

**Comment by kbecciv** [comment]
@alexpensify Issue is occurring on only IOS native, updated the step 1.

**Comment by alexpensify** [comment]
Thanks for confirming @kbecciv - I'll test using iOS native.

**Comment by alexpensify** [comment]
I'm having trouble with BrowserStack, I'll try testing again soon.

**Comment by alexpensify** [comment]
Tom helped me test on the native app and was able to replicate it.

**Comment by khaterdev** [comment]
# Proposal

Contributor details
Your Expensify account email: mostafa@mostafakhater.com
Upwork Profile Link: https://www.upwork.com/freelancers/~014ecc4269a1dbeca0

## Please re-state the problem that we are trying to solve in this issue.
If the multiline input has a fixed height (because of the screen size) and the lines are more than what can be visible on the screen, when scrolling up to the first line and setting the cursor to the end of the line then pressing Enter for a new line, the input component will scroll down and hide the first line and the new line.


https://github.com/Expensify/App/assets/10665376/b52e1f95-c9b0-4ef5-b396-7bde8936b21e



## What is the root cause of that problem?
Rerendering the text input component by changing the text value to the state draft value makes unintended changes to the selection object. [Here](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionItemMessageEdit.js#L320)

It replaces the value of the input which leads to scrolling to the end of text. This changes the selection value 3 times very quickly while having the second position value set to the end of the draft, which makes the text input component scroll/jump to catch up with the selection position changes and ends up scrolling to a random offset.

## What changes do you think we should make in order to solve the problem?
You should stop updating the text input value of the text input using the draft value saved in the component state. Instead, you can set the initial draftMessage as private property in the class constructor and replace the value prop of the Composer with defaultValue setting it to the initial draftMessage and it should just work fine.


https://github.com/Expensify/App/assets/10665376/78dedb6c-04e0-411a-8409-eb3ba2618b22



## What alternative solutions did you explore? (Optional)
N/A

**Comment by khaterdev** [comment]
# Proposal

## Please re-state the problem that we are trying to solve in this issue.
If the multiline input has a fixed height (because of the screen size) and the lines are more than what can be visible on the screen, when scrolling up to the first line and setting the cursor to the end of the line then pressing Enter for a new line, the input component will scroll down and hide the first line and the new line.


https://github.com/Expensify/App/assets/10665376/b52e1f95-c9b0-4ef5-b396-7bde8936b21e



## What is the root cause of that problem?
Rerendering the text input component by changing the text value to the state draft value makes unintended changes to the selection object. [Here](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionItemMessageEdit.js#L320)

It replaces the value of the input which leads to scrolling to the end of text. This changes the selection value 3 times very quickly while having the second position value set to the end of the draft, which makes the text input component scroll/jump to catch up with the selection position changes and ends up scrolling to a random offset.

## What changes do you think we should make in order to solve the problem?
You should stop updating the text input value of the text input using the draft value saved in the component state. Instead, you can set the initial draftMessage as private property in the class constructor and replace the value prop of the Composer with defaultValue setting it to the initial draftMessage and it should just work fine.


https://github.com/Expensify/App/assets/10665376/78dedb6c-04e0-411a-8409-eb3ba2618b22



## What alternative solutions did you explore? (Optional)
N/A

**Comment by alexpensify** [comment]
Waiting for proposals

**Comment by khaterdev** [comment]
# Proposal

## Please re-state the problem that we are trying to solve in this issue.

If the multiline input has a fixed height (because of the screen size) and the lines are more than what can be visible on the screen, when scrolling up to the first line and setting the cursor to the end of the line then pressing Enter for a new line, the input component will scroll down and hide the first line and the new line.

https://github.com/Expensify/App/assets/10665376/b52e1f95-c9b0-4ef5-b396-7bde8936b21e

## What is the root cause of that problem?

Rerendering the text input component by changing the text value to the state draft value makes unintended changes to the selection object. [(Here)](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionItemMessageEdit.js#L320) 

It replaces the value of the input which leads to scrolling to the end of text. This changes the selection value 3 times very quickly while having the second position value set to the end of the draft, which makes the text input component scroll/jump to catch up with the selection position changes and ends up scrolling to a random offset.

## What changes do you think we should make in order to solve the problem?

You should stop updating the text input value of the text input using the draft value saved in the component state. Instead, you can set the initial draftMessage as private property in the class constructor and replace the value prop of the Composer with defaultValue setting it to the initial draftMessage and it should just work fine.

https://github.com/Expensify/App/assets/10665376/78dedb6c-04e0-411a-8409-eb3ba2618b22

## What alternative solutions did you explore? (Optional)

N/A

**Comment by alexpensify** [comment]
@eVoloshchak - when you get a chance, please review the recent proposal. Thank you!

**Comment by eVoloshchak** [comment]
Huh, that is an interesting issue
Thanks for the bump, I did miss the one on the weekend!

@khaterdev's proposal looks promising, but I'm not 100% sure just yet and wanna do some testing. I'll get back with a review tomorrow

**Comment by khaterdev** [comment]
> 
> I don't think `triggerSaveOrCancel` has anything to do with the issue. It shouldn't be called when you press 'enter' key on iOS keyboard.
> 
That was a type, I meant [(this)](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionItemMessageEdit.js#L321) line.

>
> Can confirm this issue happening. The updateDraft function is called twice. The main question is: **Why** is this happening? What's causing this? Why can't we use state as we currently are? Would this still be reproducible if we migrate this to a functional component? In short, I think we should figure out what exactly is the root cause of this
>
I tested it with a new blank application. Also, used functional components. It does have the same behavior. So, it is not related to any of the Expensify codes.

Maybe it is a bug in the TextInput component itself or intended behavior. But, it looks like it is rendering the component on state change by setting the value prop to the state value, then this, somehow, might be triggering the onChangeText event, which in turn tries to set the state again [(Here)](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionItemMessageEdit.js#L319). Then it re-renders and acts like replacing the text manually, which leads to putting the cursor at the end of the text. The whole thing changes the cursor position twice very quickly. The change should make the TextInput scroll and changing it twice makes it scroll to 2 different positions which ends on a random position.

**Comment by khaterdev** [comment]
> 
> I don't think `triggerSaveOrCancel` has anything to do with the issue. It shouldn't be called when you press 'enter' key on iOS keyboard.
> 
That was a type, I meant [(this)](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionItemMessageEdit.js#L321) line.

>
> Can confirm this issue happening. The updateDraft function is called twice. The main question is: **Why** is this happening? What's causing this? Why can't we use state as we currently are? Would this still be reproducible if we migrate this to a functional component? In short, I think we should figure out what exactly is the root cause of this
>
I tested it with a new blank application. Also, used functional components. It does have the same behavior. So, it is not related to any of the Expensify codes.

Maybe it is a bug in the TextInput component itself or intended behavior. But, it looks like it is rendering the component on state change by setting the value prop to the state value, then, somehow, it might be triggering the onChangeText event, which in turn tries to set the state again [(Here)](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionItemMessageEdit.js#L319). Then it re-renders and acts like replacing the text manually, which leads to putting the cursor at the end of the text. The whole thing changes the cursor position twice very quickly. The change should make the TextInput scroll and changing it twice makes it scroll to 2 different positions which ends on a random position.

**Comment by khaterdev** [comment]
Looks like the [official React Native documentation](https://reactnative.dev/docs/handling-text-input) site is already showing it the same way I suggested in my proposal. It uses the defaultValue prop to set the initial value and the onChangeText to update the state. It is not showing updating the value of the TextInput component using the state again as it will lead to re-rending the component.

Looks like the [official React Native documentation](https://reactnative.dev/docs/handling-text-input) site is already showing it the same way I suggested in my proposal. It uses the defaultValue prop to set the initial value and the onChangeText to update the state. It is not showing updating the value of the TextInput component using the state again as it will lead to re-rending the component.

**Comment by khaterdev** [comment]
> 
> I don't think `triggerSaveOrCancel` has anything to do with the issue. It shouldn't be called when you press 'enter' key on iOS keyboard.
> 
That was a type, I meant [(this)](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionItemMessageEdit.js#L321) line.

>
> Can confirm this issue happening. The updateDraft function is called twice. The main question is: **Why** is this happening? What's causing this? Why can't we use state as we currently are? Would this still be reproducible if we migrate this to a functional component? In short, I think we should figure out what exactly is the root cause of this
>
I tested it with a new blank application. Also, used functional components. It does have the same behavior. So, it is not related to any of the Expensify codes.

Maybe it is a bug in the TextInput component itself or intended behavior. But, it looks like it is rendering the component on state change by setting the value prop to the state value, then, somehow, it might be triggering the onChangeText event, which in turn tries to set the state again [(Here)](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionItemMessageEdit.js#L319). Then it re-renders and acts like replacing the text manually, which leads to putting the cursor at the end of the text. The whole thing changes the cursor position twice very quickly. The change should make the TextInput scroll and changing it twice makes it scroll to 2 different positions which ends on a random position.


Looks like the [official React Native documentation](https://reactnative.dev/docs/handling-text-input) site is already showing it the same way I suggested in my proposal. It uses the defaultValue prop to set the initial value and the onChangeText to update the state. It is not showing updating the value of the TextInput component using the state again as it will lead to re-rending the component.

**Comment by eVoloshchak** [comment]
> Maybe it is a bug in the TextInput component itself or intended behavior.

In that case, we should fix it in `react-native` itself. 
If there's abug in react-native, this means the same issue might appear on other pages too, so an upstream fix would be better than refactoring every page. Could you share a snack with a blank application with minimal reproduction steps for this? Are there any open `react-native` issues related to this?

**Comment by khaterdev** [comment]
> > Maybe it is a bug in the TextInput component itself or intended behavior.
> 
> In that case, we should fix it in `react-native` itself. If there's a bug in react-native, this means the same issue might appear on other pages too, so an upstream fix would be better than refactoring every page. Could you share a snack with a blank application with minimal reproduction steps for this? Are there any open `react-native` issues related to this?
>

After checking again, I suspect that it might not be a bug in react-native as setting the value prop is expected to re-render the component while it was already re-rendered by changing the state using the onChangeText event. So, I see now it is already expected to re-render twice which leads to the current issue. Again, I suggest using defaultValue to set the initial value instead of setting the value prop on state change. The TextInput component already keeps the entered text without the need of resetting it every time the text changes. We just need to keep the onChangeText event to use the state value during submission.

[Here](https://snack.expo.dev/@khaterdev/expensify-chat-edit-comment-issue?platform=ios) is a snack showing two components, one using the same ideas as the current Expensify code, and the other using my suggestion. You can easily see that the issue only occurs while using the first component only.

**Comment by khaterdev** [comment]
> > Maybe it is a bug in the TextInput component itself or intended behavior.
> 
> In that case, we should fix it in `react-native` itself. If there's a bug in react-native, this means the same issue might appear on other pages too, so an upstream fix would be better than refactoring every page. Could you share a snack with a blank application with minimal reproduction steps for this? Are there any open `react-native` issues related to this?
>

After checking again, I suspect that it might not be a bug in react-native as setting the value prop is expected to re-render the component while it was already re-rendered by changing the state using the onChangeText event. So, I see now it is already expected to re-render twice which leads to the current issue. Again, I suggest using defaultValue to set the initial value instead of setting the value prop on state change. The TextInput component already keeps the entered text without the need of resetting it every time the text changes. We just need to keep the onChangeText event to use the state value during submission.

[Here](https://snack.expo.dev/@khaterdev/expensify-chat-edit-comment-issue?platform=ios) is a snack showing two components, one is setting the value prop to the state value as the current Expensify code, and the other using my suggestion by setting the defaultValue prop one time on first render. You can easily see that the issue only occurs while using the first component only.

**Comment by khaterdev** [comment]
> So knowing what we know now, how would this change your proposal, if at all, @khaterdev?
>
I don't think this would change anything. I still suggest the same solution mentioned in the proposal. We shouldn't update the value prop using the state. No need to do that. Only the state should be updated if the text is changed so we get only 1 re-render, and pass the initial value to the component using the defaultValue prop.

**Comment by alexpensify** [comment]
@deetergp should we update the description of this GH?

**Comment by khaterdev** [comment]
@alexpensify @deetergp @eVoloshchak, do you still have questions about this GH? Will we go forward?

**Comment by eVoloshchak** [comment]
@deetergp, new documentation also has a [controlled input](https://react.dev/reference/react-dom/components/input#controlling-an-input-with-a-state-variable) section. 
My point is there's nothing inherently wrong with using value and onChange (i.e. controlled input). Using an uncontrolled input would resolve the issue, but we aren't fixing the actual cause.

**Comment by khaterdev** [comment]
> Using an uncontrolled input would resolve the issue, but we aren't fixing the actual cause.

@eVoloshchak, isn't this trying to fix an issue you don't have? If you don't need to use a controlled input (for this specific case) and the application works perfectly without it, why are you trying to use it?
The issue seems to be existing for a new setup too as I mentioned before. I think that it is an expected behavior, but maybe it can be fixed within React Native itself.
So, trying to find a solution for the issue or even fixing React Native if it is a bug while we don't need it implemented in Expensify is just out of scope.

**Comment by khaterdev** [comment]
> Using an uncontrolled input would resolve the issue, but we aren't fixing the actual cause.

@eVoloshchak, isn't this trying to fix an issue you don't have? If you don't need to use a controlled input (for this specific case) and the application works perfectly without it, why are you trying to use it?
The issue seems to be existing for a new setup too as I mentioned before. I think that it is an expected behavior (at least the re-render part), but maybe it can be fixed within React Native itself.
So, trying to find a solution for the issue or even fixing React Native if it is a bug while we don't need it implemented in Expensify is just out of scope.

**Comment by eVoloshchak** [comment]
> If you don't need to use a controlled input (for this specific case) and the application works perfectly without it, why are you trying to use it?

Because the issue will come back in case we ever need to make this input controlled. And all of the controlled multiline inputs (that exist now or will be added in the future) will keep having the same issue

> The issue seems to be existing for a new setup too as I mentioned before. I think that it is an expected behavior (at least the re-render part), but maybe it can be fixed within React Native itself.

Fixing it in react-native itself would be the way to go I think

**Comment by fabriziobertoglio1987** [comment]
#### Reproducing the issue on iOS without react-native

<details><summary>CLICK TO OPEN SOURCE CODE</summary>
<p>

```objc
//
//  ViewController.m
//  iPhoneTestApp
//
//  Created by Fabrizio Bertoglio on 5/17/23.
//

#import "ViewController.h"

@interface RCTUITextView : UITextView<UITextViewDelegate>
@end

@implementation RCTUITextView {}

-(void)textViewDidChange:(UITextView *)textView
{
  NSLog(@"textViewDidChange : %@",textView.text);
  NSMutableDictionary<NSAttributedStringKey, id> *attributes = [NSMutableDictionary dictionaryWithCapacity:10];
  attributes[NSForegroundColorAttributeName] = [UIColor blueColor];
  NSAttributedString *attributedString = [[NSAttributedString alloc] initWithString:textView.text attributes: attributes];

  [textView setAttributedText: attributedString];
}

@end

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  // Do any additional setup after loading the view.
  RCTUITextView *textView = [[RCTUITextView alloc] initWithFrame:CGRectMake(10, 200, 300, 50)];
  textView.scrollEnabled = true;
  textView.delegate = textView;
  UIView *redBackgroundView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, self.view.frame.size.height / 2.0)];
  redBackgroundView.backgroundColor = [UIColor redColor];
  [self.view addSubview:redBackgroundView];
  [redBackgroundView addSubview:textView];
}


@end

```

</p>
</details>


https://user-images.githubusercontent.com/24992535/246601549-99f480f3-ce80-4678-9378-f71c8aa67e17.mp4

**Comment by fabriziobertoglio1987** [comment]
#### Fixing the issue on iOS without react-native

<details><summary>CLICK TO OPEN TESTS RESULTS</summary>
<p>

```objc
//
//  ViewController.m
//  iPhoneTestApp
//
//  Created by Fabrizio Bertoglio on 5/17/23.
//

#import "ViewController.h"

@interface RCTUITextView : UITextView<UITextViewDelegate>
@end

@implementation RCTUITextView {}

-(void)textViewDidChange:(UITextView *)textView
{
  // disable scroll to avoid issue #
  textView.scrollEnabled = false;
  NSLog(@"textViewDidChange : %@",textView.text);
  NSMutableDictionary<NSAttributedStringKey, id> *attributes = [NSMutableDictionary dictionaryWithCapacity:10];
  attributes[NSForegroundColorAttributeName] = [UIColor blueColor];
  NSAttributedString *attributedString = [[NSAttributedString alloc] initWithString:textView.text attributes: attributes];

  // save cursor position
  UITextRange *selectedRange = self.selectedTextRange;
  NSInteger oldTextLength = self.attributedText.string.length;

  [textView setAttributedText: attributedString];

  // restore cursor position
  if (selectedRange.empty) {
    // Maintaining a cursor position relative to the end of the old text.
    NSInteger offsetStart = [self offsetFromPosition:self.beginningOfDocument
                                          toPosition:selectedRange.start];
    NSInteger offsetFromEnd = oldTextLength - offsetStart;
    NSInteger newOffset = textView.text.length - offsetFromEnd;
    UITextPosition *position = [self positionFromPosition:self.beginningOfDocument
                                                   offset:newOffset];
    [self setSelectedTextRange:[self textRangeFromPosition:position toPosition:position]];
  }

  textView.scrollEnabled = true;
}

@end

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  // Do any additional setup after loading the view.
  RCTUITextView *textView = [[RCTUITextView alloc] initWithFrame:CGRectMake(10, 200, 300, 50)];
  textView.scrollEnabled = true;
  textView.delegate = textView;
  UIView *redBackgroundView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, self.view.frame.size.height / 2.0)];
  redBackgroundView.backgroundColor = [UIColor redColor];
  [self.view addSubview:redBackgroundView];
  [redBackgroundView addSubview:textView];
}


@end

```
</p>
</details>


https://user-images.githubusercontent.com/24992535/246606574-92da5805-e6b1-4675-b1e7-e97dd494a9bb.mp4

**Comment by fabriziobertoglio1987** [comment]
#### Reproducing the issue in React-Native

<details><summary>CLICK TO OPEN SOURCECODE</summary>
<p>

```objc
- (void)setAttributedText:(NSAttributedString *)attributedText
{
  [super setAttributedText:attributedText];
  [self textDidChange];
}
```

</p>
</details>


https://user-images.githubusercontent.com/24992535/246601996-fe08ce4f-6a6f-460d-bb06-da3dd0feaaaf.mp4

**Comment by fabriziobertoglio1987** [comment]
#### Fixing the issue in React-Native

<details><summary>CLICK TO OPEN SOURCECODE</summary>
<p>

```objc
- (void)setAttributedText:(NSAttributedString *)attributedText
{
  self.scrollEnabled = false;
  [super setAttributedText:attributedText];
  [self textDidChange];
  self.scrollEnabled = true;
}
```

</p>
</details>


https://user-images.githubusercontent.com/24992535/246602165-c1bc1b44-087f-4ec9-bc44-14f961eaa4d1.mp4

**Comment by fabriziobertoglio1987** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Multiline TextInput with a fixed height will scroll to the bottom of the screen when prepending new lines to the text.

### What is the root cause of that problem?

The issue is caused by iOS UITextView:
- The cursor moves to the end of the text when prepending new lines. 
- Moving the cursor to the end of the text triggers the scroll to the bottom.

The behavior was reproduced on an iOS App (without react-native). 
The example included below implements a Component RCTUITextView based on UITextView, which modifies the UITextView attributedText with the textViewDidChange callback (source code available in this [comment](https://github.com/Expensify/App/issues/19507#issuecomment-1595730348)). 

Adding a new line on top of the UITextView on iOS results in:
Issue 1) The cursor moves to the end of TextInput text
Issue 2) The TextInput scrolls to the bottom

<details><summary>Reproducing the issue on an iOS App without react-native</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/246601549-99f480f3-ce80-4678-9378-f71c8aa67e17.mp4" width="900" />

</p>
</details>


Issue 1) is already fixed in react-native, which restores the previous cursor position (on Fabric with  [_setAttributedString](https://github.com/fabriziobertoglio1987/react-native/blob/71e7bbbc2cf21abacf7009e300f5bba737e20d17/packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm#L600-L610)) after changing the text.
Issue 2) needs to be fixed in react-native.

### What changes do you think we should make in order to solve the problem?

The functionality was fixed on the iOS example with the following steps (source code available in this [comment](https://github.com/Expensify/App/issues/19507#issuecomment-1595730432)):
Issue 1) implementing logic to restore the previous cursor position as done in React Native Fabric [_setAttributedString](https://github.com/fabriziobertoglio1987/react-native/blob/71e7bbbc2cf21abacf7009e300f5bba737e20d17/packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm#L600-L610)
Issue 2) disable scrollingEnabled when changing the attributedText

<details><summary>Fixing the issue on iOS without react-native</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/246606574-92da5805-e6b1-4675-b1e7-e97dd494a9bb.mp4" width="1000" />

</p>
</details>

The solution was tested in react-native and fixed the issue as documented below.

Issue 1) already fixed
Issue 2) fixed by disabling scrollingEnabled while changing the attributedText

<details><summary>Reproducing the issue in React-Native</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/246601996-fe08ce4f-6a6f-460d-bb06-da3dd0feaaaf.mp4" width="1000" />


</p>
</details>

<details><summary>Fixing the issue in React-Native</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/246602165-c1bc1b44-087f-4ec9-bc44-14f961eaa4d1.mp4" width="1000" />


</p>
</details>

### What alternative solutions did you explore? (Optional)

I did not investigate an alternative solution.

**Comment by alexpensify** [comment]
@eVoloshchak - when you get a chance, can you review the recent proposals? Thank you!

**Comment by eVoloshchak** [comment]
@fabriziobertoglio1987, wow, thanks for such a detailed explanation!

> 1) implementing logic to restore the previous cursor position as done in React Native Fabric 

We're in the progress of migrating to Fabric, so if this part is already resolved there - that's awesome

> 2) fixed by disabling scrollingEnabled while changing the attributedText

I've tested and it does resolve the issue. However, we still need to get to the bottom of **why** the issue is happening. According to your comment, this is a bug with iOS itself. Could you please take a look if there are any issues/bug reports for iOS on this? 
I found https://github.com/facebook/react-native/issues/30207, which I think is related

**Comment by alexpensify** [comment]
Thank you @eVoloshchak! @fabriziobertoglio1987 - keep us posted on your findings or questions to the feedback provided in the last comment.

**Comment by fabriziobertoglio1987** [comment]
Thanks a lot. I further investigated this issue and this is my update.

>According to your comment, this is a bug with iOS itself. Could you please take a look if there are any issues/bug reports for iOS on this?

Commit [Maintain cursor position when the text changes in the text input](https://github.com/facebook/react-native/commit/b0e39b2ed9b66b378eb75bee9e692fc801431ddd) was introduced to fix iOS issue 652653.

https://developer.apple.com/forums/thread/652653
https://stackoverflow.com/questions/34914948/how-to-stop-cursor-changing-position-when-i-setattributedtext-in-uitextview-dele

1) Updating the UITextView attributedText causes the cursor to move to the end. 
2) Moving the cursor to the end of the text triggers a scroll to the bottom.

> However, we still need to get to the bottom of why the issue is happening. 

@eVoloshchak is correct, the root cause of this issue is in react-native itself.
- The issue is caused by adding new line character (`\n`) in the text, but does not reproduce with other characters.
- Reproduces with controlled and not controlled TextInput components in Fabric 

The root cause is:
1) [_textOf][1] fails to compare identical strings. The equality comparison between oldText `"1\n\n2\n3\n"` and newText `"1\n\n2\n3\n"` returns false.
2) Triggers an additional update of the attributedText, replacing it with an identical string. The UITextView already includes the new line, but `_textOf` fails to compare the current UITextView state (`_backedTextInputView.attributedText` with value `"1\n\n2\n3\n"`) with the next state (`attributedString` with value `"1\n\n2\n3\n"`).
3) The additional update causes the cursor to move to the end of the text (iOS issue [652653][2])
2) Moving the cursor to the end of the text triggers a scroll to the bottom.

The root cause of this issue is STEP 1) ( [_textOf][1] failed to compare identical strings).

I tested a temporary fix for the issue. I will update my proposal in a couple of days with the final solution.
The solution was only developed for testing purposes, I will update the solution soon.

[1]: https://github.com/facebook/react-native/blob/71e7bbbc2cf21abacf7009e300f5bba737e20d17/packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm#L594 "_textOf"
[2]: https://developer.apple.com/forums/thread/652653

<details><summary>Reproducing the issue on React Native not controlled TextInput</summary>
<p>

https://user-images.githubusercontent.com/24992535/247505358-4b10fc23-a0fd-4c0a-b06f-fd17bf671d77.mp4

</p>
</details>

<details><summary>Fixing the issue on React Native</summary>
<p>

<details><summary>Not Controlled Component</summary>
<p>


https://user-images.githubusercontent.com/24992535/247506063-0eaabccc-bc8f-4883-ba2f-8e5bfc1bc98e.mp4


</p>
</details> 


<details><summary>Controlled Component</summary>
<p>


https://user-images.githubusercontent.com/24992535/247507038-a95be425-3373-4d1e-bee4-bc2eb6241434.mp4



</p>
</details>

</p>
</details>

**Comment by alexpensify** [comment]
@eVoloshchak - I think we are making progress here but let me know if we should move to internal.

**Comment by eVoloshchak** [comment]
@alexpensify, yeah, there's progress being made by @fabriziobertoglio1987 
I think we should add the External and Help Wanted labels back

**Comment by alexpensify** [comment]
Done, thank you! @eVoloshchak is there any remaining feedback for @fabriziobertoglio1987 based on their latest reply?

**Comment by fabriziobertoglio1987** [comment]
I further investigated the issue. This is my update for today. 

I was able to reproduce the same issue with a bare iOS iPhone App with the following steps:

1) The iOS App renders a UITextView 
2) At initialization, the UITextView AttributedString foreground color is set to red 
3) Typing a new line triggers textDidChangeEvent
4) The UITextView foreground color changes to green on textDidChangeEvent
5) The update causes the cursor to move to the end of the string
6) Moving the cursor to the end of the string, causes the TextView to scroll to the end

The issue reproduces on the iOS App without react-native only when creating a new AttributedString, while it does not reproduce when using a mutable copy of the AttributedString.

<details><summary>Changing the foreground color of the AttributedString using a mutable copy of the original AttributedString does not scroll the TextView to the bottom</summary>
<p>

https://user-images.githubusercontent.com/24992535/248256843-55673df8-2bfe-4cfe-98cb-ed3a3bff0317.mp4

</p>
</details>

<details><summary>Creating a new AttributedString with a different foreground color causes the cursor to jump to the end of the line.</summary>
<p>

https://user-images.githubusercontent.com/24992535/248256498-a556fd65-a75a-4902-9010-f17fc88ac0da.mp4

</p>
</details>

<details><summary>Click to open sourcecode</summary>
<p>

```objc
//
//  ViewController.m
//  iPhoneTestApp
//
//  Created by Fabrizio Bertoglio on 5/17/23.
//

#import "ViewController.h"

@interface RCTUITextView : UITextView<UITextViewDelegate>
@end

@implementation RCTUITextView {
  NSAttributedString *_oldAttributedString;
  NSMutableDictionary<NSAttributedStringKey, id> *_attributes;
}

-(void)textViewDidChange:(UITextView *)textView
{
 /*
  NSMutableDictionary<NSAttributedStringKey, id> *attributes = [NSMutableDictionary dictionaryWithCapacity:10];
  attributes[NSForegroundColorAttributeName] = [UIColor greenColor];
  NSAttributedString *newAttributes = [[NSAttributedString alloc] initWithString:textView.text attributes:attributes];
  [textView setAttributedText: newAttributes];
 */

  NSMutableAttributedString *attributedStringCopy = [textView.attributedText mutableCopy];
  [attributedStringCopy addAttributes:@{ NSForegroundColorAttributeName : [UIColor greenColor]} range:NSMakeRange(0, attributedStringCopy.length)];
  [textView setAttributedText: attributedStringCopy];
}

@end

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  // Do any additional setup after loading the view.
  RCTUITextView *textView = [[RCTUITextView alloc] initWithFrame:CGRectMake(10, 200, 300, 50)];
  textView.scrollEnabled = true;
  textView.delegate = textView;

  NSMutableDictionary<NSAttributedStringKey, id> *attributes = [NSMutableDictionary dictionaryWithCapacity:10];
  attributes[NSForegroundColorAttributeName] = [UIColor redColor];
  NSAttributedString *attributedString = [[NSAttributedString alloc] initWithString:textView.text attributes: attributes];
  [textView setAttributedText:attributedString];

  UIView *redBackgroundView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, self.view.frame.size.height / 2.0)];
  redBackgroundView.backgroundColor = [UIColor redColor];
  [self.view addSubview:redBackgroundView];
  [redBackgroundView addSubview:textView];
}


@end
```

</p>
</details>

I will further work on this issue and update the proposal soon. Thanks a lot for your attention.

**Comment by alexpensify** [comment]
Thank you for that update @fabriziobertoglio1987

**Comment by alexpensify** [comment]
@eVoloshchak - could you review again for feedback? Thank you!

**Comment by eVoloshchak** [comment]
Will post a review tomorrow, this one is an interesting issue to research (Apple bugs are **_fun_**), thanks for the investigation @fabriziobertoglio1987

**Comment by alexpensify** [comment]
@eVoloshchak - any update here on the review status? Thank you!

**Comment by eVoloshchak** [comment]
Can reproduce this on a bare project (without react-native), thank you @fabriziobertoglio1987 for those examples

1. This is a bug in iOS (is there a ticket on Apple's tracker for this?)
2. This was partially fixed by https://github.com/facebook/react-native/commit/b0e39b2ed9b66b378eb75bee9e692fc801431ddd in Fabric
3. Disabling scrolling when changing the `attributedText` fixes the issue, but this is a workaround
4. I'm fairly certain Apple won't fix this any time soon

We should avoid workarounds and fix the root cause instead. On the other hand, this is a bug with iOS itself which we can't fix, and Apple won't fix it probably. However, in my opinion, the bug isn't severe enough to justify the use of a workaround solution right away. 
I think we should wait and seek a better solution, only if there is no other option we should apply a workaround
Either way, a good place to start would be to 

- open a ticket for this bug on Apple's forums (if there isn't one already)
- open an issue for this in `react-native `repo (if there isn't one already)

**Comment by alexpensify** [comment]
Reassigning another `bug` team member, I'm going OOO until Monday, July 10, and will take it back then if it's not closed by then. 

Required action from the 🐛 team: Make sure the proposal process is moving forward, so we can get this one assigned out to fix. Also, we might have to start a discussion if we should close.

**Comment by fabriziobertoglio1987** [comment]
#### The root cause of the Issue

- Prepending a new line with `\n` causes a change in the AttributedString
- Changing the AttributedString attributes causes the cursor to move to the end of the string
- Moving the cursor to the end of the string causes the TextView to scroll to the end

Changing any field of the AttributedString causes the scroll to the end (more info in https://github.com/Expensify/App/issues/19507#issuecomment-1607559935). 

Prepending `\n` triggers an update of the AttributedString attributes which applies the following changes:

1) old AttributedString has 8 blocks, string `"1\n\n2"`, and a total length of 5. 3 blocks (spans) are used and have the following styles.

```
Printing description of oldAttributedString->attributes:
Length 5 (8 blocks, 3 used, block 0 is at 0)
 2 0x600002e70230 {
    NSBackgroundColor = "UIExtendedGrayColorSpace 0 0";
    NSColor = "UIExtendedGrayColorSpace 0 1";
    NSFont = "<UICTFont: 0x130457dc0> font-family: \".SFUI-Regular\"; font-weight: normal; font-style: normal; font-size: 14.00pt";
    NSParagraphStyle = "Alignment 4, LineSpacing 0, ParagraphSpacing 0, ParagraphSpacingBefore 0, HeadIndent 0, TailIndent 0, FirstLineHeadIndent 0, LineHeight 0/0, LineHeightMultiple 0, LineBreakMode 0, Tabs (\n    28L,\n    56L,\n    84L,\n    112L,\n    140L,\n    168L,\n    196L,\n    224L,\n    252L,\n    280L,\n    308L,\n    336L\n), DefaultTabInterval 0, Blocks (\n), Lists (\n), BaseWritingDirection 0, HyphenationFactor 0, TighteningForTruncation NO, HeaderLevel 0 LineBreakStrategy 0 PresentationIntents (\n) ListIntentOrdinal 0 CodeBlockIntentLanguageHint ''";
}
 1 0x600002f9d5e0 {
    NSBackgroundColor = "UIExtendedGrayColorSpace 0 0";
    NSColor = "UIExtendedGrayColorSpace 0 1";
    NSFont = "<UICTFont: 0x130457dc0> font-family: \".SFUI-Regular\"; font-weight: normal; font-style: normal; font-size: 14.00pt";
    NSParagraphStyle = "Alignment 0, LineSpacing 0, ParagraphSpacing 0, ParagraphSpacingBefore 0, HeadIndent 0, TailIndent 0, FirstLineHeadIndent 0, LineHeight 0/0, LineHeightMultiple 0, LineBreakMode 0, Tabs (\n    28L,\n    56L,\n    84L,\n    112L,\n    140L,\n    168L,\n    196L,\n    224L,\n    252L,\n    280L,\n    308L,\n    336L\n), DefaultTabInterval 0, Blocks (\n), Lists (\n), BaseWritingDirection 0, HyphenationFactor 0, TighteningForTruncation NO, HeaderLevel 0 LineBreakStrategy 0 PresentationIntents (\n) ListIntentOrdinal 0 CodeBlockIntentLanguageHint ''";
}
 2 0x600002e70230 {
    NSBackgroundColor = "UIExtendedGrayColorSpace 0 0";
    NSColor = "UIExtendedGrayColorSpace 0 1";
    NSFont = "<UICTFont: 0x130457dc0> font-family: \".SFUI-Regular\"; font-weight: normal; font-style: normal; font-size: 14.00pt";
    NSParagraphStyle = "Alignment 4, LineSpacing 0, ParagraphSpacing 0, ParagraphSpacingBefore 0, HeadIndent 0, TailIndent 0, FirstLineHeadIndent 0, LineHeight 0/0, LineHeightMultiple 0, LineBreakMode 0, Tabs (\n    28L,\n    56L,\n    84L,\n    112L,\n    140L,\n    168L,\n    196L,\n    224L,\n    252L,\n    280L,\n    308L,\n    336L\n), DefaultTabInterval 0, Blocks (\n), Lists (\n), BaseWritingDirection 0, HyphenationFactor 0, TighteningForTruncation NO, HeaderLevel 0 LineBreakStrategy 0 PresentationIntents (\n) ListIntentOrdinal 0 CodeBlockIntentLanguageHint ''";
}
```

2) The new AttributedString has 2 blocks, 1 used (1 span), string is "1\n\n2". The total length is 5 blocks.

```
Printing description of attributedString->attributes:
Length 5 (2 blocks, 1 used, block 0 is at 0)
 5 0x600002e70230 {
    NSBackgroundColor = "UIExtendedGrayColorSpace 0 0";
    NSColor = "UIExtendedGrayColorSpace 0 1";
    NSFont = "<UICTFont: 0x130457dc0> font-family: \".SFUI-Regular\"; font-weight: normal; font-style: normal; font-size: 14.00pt";
    NSParagraphStyle = "Alignment 4, LineSpacing 0, ParagraphSpacing 0, ParagraphSpacingBefore 0, HeadIndent 0, TailIndent 0, FirstLineHeadIndent 0, LineHeight 0/0, LineHeightMultiple 0, LineBreakMode 0, Tabs (\n    28L,\n    56L,\n    84L,\n    112L,\n    140L,\n    168L,\n    196L,\n    224L,\n    252L,\n    280L,\n    308L,\n    336L\n), DefaultTabInterval 0, Blocks (\n), Lists (\n), BaseWritingDirection 0, HyphenationFactor 0, TighteningForTruncation NO, HeaderLevel 0 LineBreakStrategy 0 PresentationIntents (\n) ListIntentOrdinal 0 CodeBlockIntentLanguageHint ''";
}
```

3) The old AttributedString block 1 has NSParagraphStyle Alignment 0
4) The new AttributedString block 5 has NSParagraphStyle Alignment 4

https://user-images.githubusercontent.com/24992535/250248022-27665b9c-bd23-403f-819e-a9ca438d3350.mp4

#### Alternative Solution - Avoid updating AttributedString without relevant changes in the text or attributes

The alternative solution is detecting this scenario based on the following conditions:
1) The two strings are identical (the old string is "1\n\n2" and the new string is "1\n\n2")
2) No relevant changes in attributes (in the case above all attributes are identical, but they are applied using 3 spans instead of one span).

As there are no relevant changes in the attributes or in the string, avoid updating the NSAttributedString, which causes the TextView to scroll to the end (https://github.com/Expensify/App/issues/19507#issuecomment-1600545816).

**Comment by MitchExpensify** [comment]
@eVoloshchak how does @fabriziobertoglio1987 's [updated proposal](https://github.com/Expensify/App/issues/19507#issuecomment-1615584137) look to you?

**Comment by eVoloshchak** [comment]
> Prepending a new line with \n causes a change in the AttributedString
Changing the AttributedString attributes causes the cursor to move to the end of the string
Moving the cursor to the end of the string causes the TextView to scroll to the end

@fabriziobertoglio1987, awesome! Is there an issue for this on Apple's forums? Or in react-native repository?

> The alternative solution is detecting this scenario based on the following conditions:
The two strings are identical (the old string is "1\n\n2" and the new string is "1\n\n2")
No relevant changes in attributes (in the case above all attributes are identical, but they are applied using 3 spans instead of one span).

This is a workaround. I suspect Apple won't fix this, so we'll probably end up implementing a workaround as a last resort. But before that, I think we should try and find if there is any discussion/bug reports on this

**Comment by deetergp** [comment]
Were you able to find any open issues in the Apple forums, @fabriziobertoglio1987?

**Comment by alexpensify** [comment]
Alright, removing the other 🐛 team member. I'm back online and see that this is our current state: 

Waiting on a reply to @deetergp's question

**Comment by fabriziobertoglio1987** [comment]
>Were you able to find any open issues in the Apple forums

Thanks @deetergp. I'm currently working on other 2 issues, sorry for my quick reply.

1) The Apple issue tracker is not public, bug reports are filed privately to Apple https://developer.apple.com/forums/thread/712889
2) There is an open [issue](https://developer.apple.com/forums/thread/652653) in the Apple Forums and the solution is in this StackOverflow [discussion](https://stackoverflow.com/questions/34914948/how-to-stop-cursor-changing-position-when-i-setattributedtext-in-uitextview-dele). The solution was implemented in react-native with commit https://github.com/facebook/react-native/commit/b0e39b2ed9b66b378eb75bee9e692fc801431ddd (more info in this [comment](https://github.com/Expensify/App/issues/19507#issuecomment-1607559935)).

@eVoloshchak https://github.com/Expensify/App/issues/19507#issuecomment-1597488640
>I found https://github.com/facebook/react-native/issues/30207, which I think is related

https://developer.apple.com/documentation/uikit/uiscrollview?language=objc
>UIScrollView is the superclass of several UIKit classes, including [UITableView](https://developer.apple.com/documentation/uikit/uitableview?language=objc) and [UITextView](https://developer.apple.com/documentation/uikit/uitextview?language=objc).

I found these reports on the Apple forums, but did not yet have the time to reproduce and debug those issues on iOS.
- https://developer.apple.com/forums/thread/691446
- https://developer.apple.com/forums/thread/662056

**Comment by alexpensify** [comment]
Thank you for this update @fabriziobertoglio1987!

**Comment by deetergp** [comment]
So basically they have implemented the workaround from the Apple forum issue in React Native proper? Nice! 

Let me restate what you said in your previous comment to make sure I am understanding it correctly. In order to fix this bug (whose bug status is debatable since it's a default behavior for the platform) we need to: 
* If we implement a version of react-native that contains the linked commit, it should take care of re-positioning the cursor when the text has NOT changed.
* In the instances where the text has changed, we would need to implement a workaround where we disable scrolling (just for IOS).

But, according to our README's [Philosophy](https://github.com/Expensify/App#philosophy) section 5.ii:
> If the reason you can't write cross-platform code is because there is a bug in ReactNative that is preventing it from working, the correct action is to fix RN and submit a PR upstream -- not to hack around RN bugs with platform-specific code paths.

It sounds like we should be trying to implement this workaround upstream in react-native.

**Comment by fabriziobertoglio1987** [comment]
Thanks @deetergp 

>It sounds like we should be trying to implement this workaround upstream in react-native.

My solution is implementing the fix upstream in react-native ([proposal](https://github.com/Expensify/App/issues/19507#issuecomment-1595731324)). I reproduced the issue in facebook/react-native and tested the fix in the videos below. I will update my proposal soon, to make it more clear.

| Before applying the fix to facebook/react-native    |
| ----------- |
| <video src="https://user-images.githubusercontent.com/24992535/246601996-fe08ce4f-6a6f-460d-bb06-da3dd0feaaaf.mp4" width="1000" />      |

| After applying the fix to facebook/react-native    |
| ----------- |
| <video src="https://user-images.githubusercontent.com/24992535/246602165-c1bc1b44-087f-4ec9-bc44-14f961eaa4d1.mp4" width="1000" />      |

**Comment by alexpensify** [comment]
Thank you for that update!

**Comment by alexpensify** [comment]
@deetergp could you review the latest update here, thanks

**Comment by alexpensify** [comment]
@deetergp any update here?

**Comment by alexpensify** [comment]
@deetergp is OOO this week. I'm going to see if I can get another engineer to review it.

**Comment by alexpensify** [comment]
I ran out of time yesterday, I'll shop around on Monday.

**Comment by alexpensify** [comment]
@dangrous - @deetergp was OOO and working limited hours this week. We need an engineer to review @fabriziobertoglio1987 feedback here: https://github.com/Expensify/App/issues/19507#issuecomment-1635281369. Thanks!

**Comment by dangrous** [comment]
Hi! Reading back through this, it seems like the migration to Fabric as well as the (pretty minor) change @fabriziobertoglio1987 suggests will solve this problem. I'm not quite sure on the next steps though - @fabriziobertoglio1987 were you going to submit a PR directly to facebook/react-native? Or were we going to put it in our fork as well? I haven't dealt with an upstream RN fix yet, sorry!

**Comment by eVoloshchak** [comment]
@dangrous, I think the standard process for this would be to do both: submit a PR to `facebook/react-native` as well as put it in our fork so we can fix the bug right away

**Comment by dangrous** [comment]
k, let's do it. I'll assign you, @fabriziobertoglio1987, and you can double up the PRs for our fork and the real deal. Thank you!

**Comment by fabriziobertoglio1987** [comment]
#### Update 26/07

Thanks a lot! I'm working on another task this week. 
I would start working on this task next week. I plan to publish PR between the 14-18th of August.

**Comment by eVoloshchak** [comment]
@fabriziobertoglio1987, keep in mind that according to [CONTRIBUTING.md](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#payment-for-contributions)

> Payment timeline all based on the day and timestamp the contributor has an accepted proposal and is assigned to the Github issue
Merged PR within 3 business days (72 hours) - 50% bonus
Merged PR within 6 business days (144 hours) - 0% bonus
Merged PR within 9 business days (216 hours) - 50% penalty
No PR within 12 business days - Contract terminated

I think this applies only to the PR in our fork of `react-native`, since PR in `facebook/react-native` is out of our control

**Comment by dangrous** [comment]
Yeah we'd still like to stick to that timeline for our fork - do you think there's a way you could get to it sooner, @fabriziobertoglio1987? If not we should probably go ahead and open this one back up, since as @eVoloshchak noted 8/14-8/18 will be outside of even the 12 day timeline.

Thanks!

**Comment by fabriziobertoglio1987** [comment]
#### Update 27/07

Sorry @dangrous @eVoloshchak. Yes. I can publish the PR earlier. 
I would publish the PR between the 2nd and the 9th of August. Would that be okay? Thanks

**Comment by dangrous** [comment]
Hi @fabriziobertoglio1987, thanks for the update! Given that that is still pretty far in the future, we're going to go ahead and unassign you, in case there's someone available who'd like to give this issue a shot in the meantime - but if the issue is still available once you're free to work on it, feel free to ping and we can get you assigned again!

**Comment by alexpensify** [comment]
Accepting proposals again here

**Comment by fabriziobertoglio1987** [comment]
@dangrous @alexpensify I already accepted the upwork job. I will publish the PR today. Thanks a lot

**Comment by fabriziobertoglio1987** [comment]
#### Update 29/07

Published PRs https://github.com/facebook/react-native/pull/38679 and https://github.com/Expensify/react-native/pull/60.

**Comment by dangrous** [comment]
Oh, amazing! @eVoloshchak do you want to give the PR in our fork a look (I can't assign you directly) and then I can check it out as well?

@alexpensify let me know if anything weird will happen if I reassign @fabriziobertoglio1987

**Comment by alexpensify** [comment]
@fabriziobertoglio1987 was assigned by me over the weekend.

**Comment by dangrous** [comment]
Oh duh, I should have looked more closely at the sidebar haha

**Comment by alexpensify** [comment]
All good! 

Looks like the PRs are moving along!

**Comment by dangrous** [comment]
PR is merged into our fork, though I've learned we're getting rid of our fork haha.

We'll just have to wait for FB to merge the upstream PR, which is out of our hands. We should track the timing based off of this PR though!

**Comment by dangrous** [comment]
Gonna switch this to weekly until we hear about the other PR, since there's literally nothing any of us can do.

**Comment by alexpensify** [comment]
Weekly update: On hold for the FB update

**Comment by alexpensify** [comment]
Weekly update: On hold for the FB update, which is moving

**Comment by fabriziobertoglio1987** [comment]
### Update 19/08

Completed [changes](https://github.com/facebook/react-native/pull/38679) requested in the code review. I will move the PR to ready for review on Monday 21st of August.

https://github.com/facebook/react-native/assets/24992535/14811724-8dc7-40f9-824e-713ff14c45df

**Comment by fabriziobertoglio1987** [comment]
### Update 20/08

I moved the PR to ready to review.

**Comment by alexpensify** [comment]
Weekly Update, it looks like the FB update is moving along and almost to the finish line.

**Comment by alexpensify** [comment]
Weekly Update: It looks like the FB PR is blocked. Is there any action we can take to unblock it or continue to hold?

**Comment by fabriziobertoglio1987** [comment]
### Update 31/08

I contacted the maintainer and I will collaborate with him until the PR is merged (relevant [comment](https://github.com/facebook/react-native/pull/38679#issuecomment-1700542491)).

**Comment by dangrous** [comment]
Thanks for that @fabriziobertoglio1987! Looks like we're close, based on the reviewer's last comment

**Comment by alexpensify** [comment]
Weekly update: FB is really close!

**Comment by dangrous** [comment]
@fabriziobertoglio1987 do you maybe want to check in with the reviewer? No worries if you don't feel comfortable, but it's been a bit of time since their latest update. Thanks!

**Comment by fabriziobertoglio1987** [comment]
Thanks @dangrous. No problem. I'll reach out and follow up. I'll be available 2-3 days a week from the 18th of September till the 18th of October because I'm renovating my new home.

**Comment by alexpensify** [comment]
Weekly update: Waiting for FB to respond 

Good luck with the renovations @fabriziobertoglio1987!

**Comment by alexpensify** [comment]
Weekly update: Waiting for FB to bring in some specialist to review

**Comment by eVoloshchak** [comment]
Same as above, bu updates in https://github.com/facebook/react-native/pull/38679 yet

**Comment by alexpensify** [comment]
Thank you @eVoloshchak for the update

**Comment by alexpensify** [comment]
Weekly update: Waiting for FB

**Comment by alexpensify** [comment]
Weekly update: Waiting for FB

**Comment by dangrous** [comment]
@fabriziobertoglio1987 do you mind reaching out to FB on that issue again, to see if we can get it moving?

**Comment by fabriziobertoglio1987** [comment]
Thanks a lot @dangrous . I contacted the FB Team (https://github.com/facebook/react-native/pull/38679#issuecomment-1774322090) and I will help them to merge the PR in the main branch. 

Now I'm waiting for their feedback. I will be again available from the 28th of October.

**Comment by alexpensify** [comment]
Thank you @eVoloshchak for the update!

**Comment by eVoloshchak** [comment]
No updates, still waiting for FB

**Comment by fabriziobertoglio1987** [comment]
I'm back and I'm happy to help get the PR merged into the main branch. Thanks

**Comment by alexpensify** [comment]
Thanks! Keep us posted @fabriziobertoglio1987.

**Comment by alexpensify** [comment]
Weekly Update: Still on hold because of FB

**Comment by alexpensify** [comment]
Weekly Update: Still on hold because of FB

**Comment by alexpensify** [comment]
Weekly Update: Still on hold because of FB

**Comment by alexpensify** [comment]
Weekly Update: Still on hold because of FB

**Comment by alexpensify** [comment]
Weekly Update: Still on hold because of FB

**Comment by fabriziobertoglio1987** [comment]
### Update 17/12/2023

- Updated branch based on the latest main and tested functionalities
- I will request code review from the react-native team during the working weekdays

Thanks

**Comment by alexpensify** [comment]
Flagging that I will be OOO until January 4. I'll check then for a weekly review.

**Comment by dangrous** [comment]
yeah we're still waiting on FB here

**Comment by dangrous** [comment]
Still no response from FB, so I guess we just chill here...

**Comment by dangrous** [comment]
Hey @roryabraham this is reliant on [an upstream PR](https://github.com/facebook/react-native/pull/38679) to the react-native repository... Is there anything we (or one of our expert contributor agencies) can do to get this moving along? It's literally been months of nothing, unfortunately.

**Comment by alexpensify** [comment]
@dangrous good call to ask if we get our partners involved here. Thank you!

**Comment by fabOnReact** [comment]
I can include the patch in the https://github.com/fabOnReact/react-native-improved library. 
I patch the native iOS/Android ReactTextView/ReactEditText APIs, but without building from source.

There will be no issues upgrading. My library creates new native APIs based on ReactTextView, ReactEditText etc...  
I don't modify JS/Objc/Java files with patch-package. 

It does not require building from source react-native and the build will not be slow.

It is the suggested approach from the react-native core team (see this [discussion](https://github.com/react-native-community/discussions-and-proposals/discussions/731#discussioncomment-7651145)):

- The Components are based on ReactTextView/ReactEditText/ReactScrollView.
- The library includes patch fixes for react-native and third-party libraries. 
- The patch are applied only until PR is merged, otherwise they will be automatically removed.

The advantages are:
- Increased engagement with the community, issues are quickly fixed (days instead of years)
- No limitations when merging PRs (react-native main branch is used on facebook marketplace).
- Allows us to further experiment and develop react-native.

Thanks a lot 🙏

**Comment by alexpensify** [comment]
@dangrous can we get your feedback on the recent update? Thanks!

**Comment by dangrous** [comment]
Hey @fabOnReact - thanks for the suggestion! I've bothered Callstack about getting the existing PR merged to react native, and they should be able to hopefully handle this week - especially since that lint error seems to be fixed now. I think we can stick to that library for now rather than switching to a new one. Stay tuned!

**Comment by alexpensify** [comment]
@dangrous any luck with the update from our partner? Thanks!

**Comment by dangrous** [comment]
Hrm, let me see if I can get an ETA! They're currently prioritizing other things but they said they'd reach out to Nick, the reviewer

**Comment by dangrous** [comment]
Callstack's attempting to help! https://github.com/facebook/react-native/pull/38679#issuecomment-1920864833 I think they're also backchanneling to see if another RN reviewer can take over from Nick

**Comment by alexpensify** [comment]
Weekly Update: #Progress! Thanks for the news!

**Comment by dangrous** [comment]
Finally got some follow up questions from FB on the PR, @fabOnReact!

**Comment by alexpensify** [comment]
Awesome, thanks @dangrous for this weekly update. @fabOnReact keep us posted if you need more time to review the questions.

**Comment by fabOnReact** [comment]
Thanks. I look into it this weekend.

**Comment by alexpensify** [comment]
Weekly Update: Waiting for feedback to the PR comments

**Comment by alexpensify** [comment]
Weekly Update: There was movement in the PR last week and it looks like there are some required tests being prepared to keep moving forward.

**Comment by alexpensify** [comment]
Weekly Update: Still waiting on the tests

**Comment by deetergp** [comment]
Weekly Update: There does not appear to be much movement on the PR since the last weekly update.

**Comment by deetergp** [comment]
It looks like the PR we've been holding on was merged last week 🎉

**Comment by alexpensify** [comment]
@deetergp or @dangrous - following this FB news, are there any more actions here to complete?

**Comment by dangrous** [comment]
So it's been merged into `main` on react-native, but we still have to wait for it to actually make it into a release - we can check the tags at the bottom of the commit message [here](https://github.com/facebook/react-native/commit/f6badca2f9a5f1e986dd76444bebde0d6049513d). Once that's done, we need to wait (again, haha) until we start using that react-native version in our app. Unfortunately I think there's still a way to go.

Anyone who knows better, though - let me know if I've got that wrong!

**Comment by deetergp** [comment]
Should we demote this to monthly since "Hurry up and wait" continues to be our next step?

**Comment by dangrous** [comment]
Still waiting.

**Comment by fabOnReact** [comment]
#### Update 5 June 2024
I will look into this next week. Thanks

**Comment by dangrous** [comment]
Yeah I'm not seeing this in any release yet, sigh, might be good to check in!

**Comment by dangrous** [comment]
@roryabraham I was wondering - do you have any thoughts on if there's a way to speed this up, or otherwise close out this issue? Short version - @fabOnReact opened a PR to react-native that was merged successfully (or I mean copied and merged, due to their weird process) but it's not in a release yet. Should we create a patch in the meantime? I'm not sure how that process works, I figured you'd be a good person to ask!

**Comment by fabOnReact** [comment]
I will ask for a cherry pick tomorrow in the upcoming release. Thanks

**Comment by fabOnReact** [comment]
#### Update 13 June 2024

- Cherry Pick Request for 0.74 https://github.com/reactwg/react-native-releases/issues/314
- Status Table for Cherry Pick Requests for 0.74 https://github.com/orgs/reactwg/projects/5/views/3
- Instructions on how to do Cherry Pick Requests https://github.com/reactwg/react-native-releases/discussions/117
- Expensify Upgrade to 0.74 discussion https://github.com/Expensify/App/issues/37374

**Comment by alexpensify** [comment]
Thank you, @fabOnReact, for the update!

**Comment by dangrous** [comment]
okay great! looks like we'll be in 0.75. Progress!

<img width="152" alt="image" src="https://github.com/Expensify/App/assets/5487802/c80cf7b6-62a8-453a-a4c5-3836c28071d8">

**Comment by fabOnReact** [comment]
The commit was cherry picked in 0.74. thanks

**Comment by fabOnReact** [comment]
> #### Update 13 June 2024
> 
> cc @dangrous 
> 
> - Cherry Pick Request for 0.74 https://github.com/reactwg/react-native-releases/issues/314
> - Status Table for Cherry Pick Requests for 0.74 https://github.com/orgs/reactwg/projects/5/views/3
> - Instructions on how to do Cherry Pick Requests https://github.com/reactwg/react-native-releases/discussions/117
> - Expensify Upgrade to 0.74 discussion https://github.com/Expensify/App/issues/37374

The commit was cherry picked in 0.74. Thanks

**Comment by dangrous** [comment]
Oh okay awesome! So basically we're just waiting for our own upgrade to RN 0.74?

**Comment by fabOnReact** [comment]
@dangrous Yes. It was picked in 0.74.3. https://github.com/orgs/reactwg/projects/5/views/3?filterQuery=Multiline

https://github.com/facebook/react-native/blob/0a683e395a72111c35f162d956331634673559e5/packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm#L610

**Comment by alexpensify** [comment]
Monthly Update: I think we are still waiting for 0.74

**Comment by dangrous** [comment]
Looks like the 0.74 upgrade is nearly ready to merge!

**Comment by dangrous** [comment]
Updated the title since we're not on hold anymore, just waiting for the merge

**Comment by dangrous** [comment]
I'm not quite sure how long it's going to be... I guess we could update to hold for https://github.com/Expensify/App/issues/37374 instead?

**Comment by dangrous** [comment]
Why is melvin talking about proposals? That's weird. I'll put it on hold for that PR to stop that from happening.

**Comment by dangrous** [comment]
oh wait it's not on staging yet so we can't test yet. But soooooon

**Comment by dangrous** [comment]
on staging I think! Let's test and close out, if we can!

**Comment by alexpensify** [comment]
Heads up, I will be offline until Tuesday, September 3, 2024, and will not actively watch over this GitHub during that period. 

If anything urgent is needed here, please ask for help in the #expensify-open-source Slack Room-- thanks! If we can wait, I can work on the payment process next week.

**Comment by fabOnReact** [comment]
I don't have a computer. My laptop is at Apple Service. Sorry

**Comment by fabOnReact** [comment]
I don't have a computer. My laptop is at Apple Service. Sorry

![IMG-20240828-WA0013.jpg](https://github.com/user-attachments/assets/2f6fa572-338c-4b1d-afc9-9a592172e928)

**Comment by dangrous** [comment]
Confirmed tested on iOS staging! We should be good (finally) for payment! @alexpensify is out until 9/3, but ready when you're back!

**Comment by alexpensify** [comment]
@dangrous, thanks for the update! Tomorrow, I'll start working on the payment process.

**Comment by alexpensify** [comment]
Payouts due: 4 September 2024

- [ ] Contributor: $1000 @fabOnReact 
- [ ] Contributor+: $1000 @eVoloshchak - please submit a request via Chat, thanks! 

Upwork job is [here](https://www.upwork.com/ab/applicants/1664050603930722304/hired). 

--- 

@fabOnReact in Upwork says this project was paid in August 2023. Can you please double-check and reply if Upworks's information is inaccurate? Thanks! For now, I'm going to close this GH.

**Comment by JmillsExpensify** [comment]
$1,000 approved for @eVoloshchak
